### ISSUE TITLE
[Bug Report] Sin and cos function results significantly deviates from the real value for certain inputs 

### ISSUE THREAD

**Issue by marty1885** [tenstorrent/tt-metal#12753]
**Describe the bug**
Recently GGML added unittests for sin and cos operators so I added them to my backend. I find that the TTNN sin/cos operator outputs significantly wrong values for certain inputs (looks like negative numbers, but I'm not certain that's the only issue).

Here's the code I came up with. It calls sin with the value -3.1415. Which we expect to output to be something close to 0. 

```c++
#include <cstddef>
#include <ttnn/operations/copy.hpp>
#include <ttnn/operations/eltwise/binary/binary.hpp>
#include <ttnn/deprecated/tt_numpy/functions.hpp>
#include <ttnn/operations/data_movement/tilize_with_val_padding/tilize_with_val_padding.hpp>

int main()
{
    auto& device = ttnn::open_device(0);
    auto t = tt::numpy::ones({1, 1, 32, 32}).to(&device);
    t = ttnn::tilize_with_zero_padding(t);
    t = ttnn::sin(ttnn::multiply(t, -3.1415));
    t.print();
}
```

However for me it prints 0.27539.

```
                 Device | INFO     | Opening user mode device driver
2024-09-17 07:55:31.250 | INFO     | SiliconDriver   - Detected 1 PCI device : [0]
                  Metal | INFO     | Initializing device 0. Program cache is NOT enabled
                  Metal | INFO     | AI CLK for device 0 is:   1000 MHz
ttnn.Tensor([[[[ 0.27539,  0.27539,  ...,  0.27539,  0.27539],
               [ 0.27539,  0.27539,  ...,  0.27539,  0.27539],
               ...,
               [ 0.27539,  0.27539,  ...,  0.27539,  0.27539],
               [ 0.27539,  0.27539,  ...,  0.27539,  0.27539]]]], shape=Shape([1, 1, 32, 32]), dtype=DataType::BFLOAT16, layout=Layout::TILE)
                  Metal | INFO     | Closing device 0
                  Metal | INFO     | Disabling and clearing program cache on device 0
                 Device | INFO     | Closing user mode device drivers
```

**To Reproduce**
1. Run the provided code (or a python equivalent)
2. See output is not expected

**Expected behavior**
Outputs the correct sine and cosine values

**Screenshots**

**Please complete the following environment information:**
- OS: Arch Linux
- Card: Grayskull e75

**Additional context**

**Comment by eyonland** [comment]
Hi @marty1885, I see that the architecture is GS here.   Does this issue also happen on WH?

**Comment by marty1885** [comment]
Hi, sorry i can't try on an WH right now. The dev box is scheduled to arrive end of month while my n300 is not connected to excessive fan noise.

**Comment by dmakoviichuk-tt** [comment]
hey @marty1885 nice catch!
Out of curiosity could you confirm that it works fine in a range [-pi/4, +pi/4] for example?
You might get very quick workaround till it get fixed.

**Comment by marty1885** [comment]
I made a plot showing the difference between TT's sin function and what C++ computes given the same `x` between -PI and PI. The red line is TTNN. It works quite well in the positive range but starts to show sings of problem from -PI/4 and really starts going bad from -PI/2.

![c1](https://github.com/user-attachments/assets/4b558318-6cf3-457b-ab2c-b07a063d580d)

**Comment by umadevimcw** [comment]
@marty1885 @eyonland Replicated the above code in Python and tested it in WHB0 N150. You can find the input and output of the TT sin function in the image below. 

Also, I have printed the output (the inputs passed in TT are given as input to the Python function as well) of the standard sin function in Python.

In WHB0 output  is similar to python

<img width="1220" alt="image" src="https://github.com/user-attachments/assets/abcc3094-25c5-426a-a067-b420b35a9019">

**Comment by umadevimcw** [comment]
Also I have tested the same code in GS where the output is same as you shared in the issue description

<img width="1234" alt="image" src="https://github.com/user-attachments/assets/46cafc12-aaf9-4d46-87ee-849f382a050e">

**Comment by umadevimcw** [comment]
The difference between GS and WHB0 implementation is that, 
WHB0 - Included range reduction technique with the help of typecast op 
Whereas in
GH - No handling of range reduction, it just conversion from [-pi, pi] to [0, 2pi] hence the difference in the output.

Also, the Tylor/Maclaurin series is used to get Sin/Cos values, to get more precision/accurate values we might need to include more factors or use another approximation approach

Please find the difference in the implementation in the image

<img width="1399" alt="image" src="https://github.com/user-attachments/assets/9fc158c3-082f-408e-871e-0322f55afadb">

**Comment by eyonland** [comment]
Great work @umadevimcw, thanks for isolating the issue.   Glad to see this does not impact WH.   

My thought is to prioritize WH bugs and the Balckhole (BH) bringup effort over the fix for this.      

@marty1885, how critical is it to get this working on GS for you?   Is this blocking your work?

**Comment by umadevimcw** [comment]
@marty1885 @eyonland Please find the plots that I have ploted on GS and WHB0 with following code 

`import numpy as np
import matplotlib.pyplot as plt
import ttnn
import torch
import math

device = ttnn.open_device(0)

x = np.linspace(-math.pi, math.pi, 100)
np_out = np.sin(x)
t_in = torch.from_numpy(x)
tensor = ttnn.from_torch(t_in, dtype=ttnn.bfloat16, layout=ttnn.TILE_LAYOUT, device=device)
t_out = ttnn.sin(tensor)
torch_output_tensor = ttnn.to_torch(t_out).type(torch.float32)

tt_out = torch_output_tensor.numpy().flatten()
# print(tt_out)
plt.plot(x, tt_out,'-b', label="TT out")
plt.plot(x, np_out, '-r', label="Standard out")
plt.legend(loc='upper center')
plt.savefig("test_line_whb0.png")
plt.close()
`
** GS **
![test_line_gs](https://github.com/user-attachments/assets/5eca26a5-462b-411e-93a4-b0ced9eff633)


** WHB0 **
![test_line_whb0](https://github.com/user-attachments/assets/e3011dd7-b2c3-47ce-a70f-ccba70c5b00b)

**Comment by umadevimcw** [comment]
@eyonland and @marty1885 Please find the dot plots as well 

### **GS**

![test_dot_gs](https://github.com/user-attachments/assets/d8682ac1-f5ee-4d71-94a0-d187859f289a)

### **WHB0**

![test_line_whb0](https://github.com/user-attachments/assets/69c39471-1eb2-4785-9f7c-c93ce82ee916)

**Comment by marty1885** [comment]
@eyonland I can just mask support for sin/cos on GS and wait for my WH to arrive. Won't be a problem in the short term. But I believe it blocks on device ROPE on GS. I'll need it fixed eventually (or drop GS support, which I'm tempted to do).

Still 2~3 weeks until my WH arrives. Let's leave it as is for now. I might send patches if I ended up with more time to play around.

**Comment by eyonland** [comment]
@umadevimcw , if we can support the typecast on GS then let's get this fix done.  If it is not available, we may need to close it given that we have higher items and given that @marty1885 will hopefully get a WH within 3 weeks.
